Migrate from PR_Poll to epoll and timerfd.

llile opened the pull request:
Now that the multi-listener, turbo and threading changes have been merged, I wanted to brush off the epoll patch again.  The prior patch had unexplained performance issues and since 389-ds-base:main has changed so much, I started over from scratch.

This version of the epoll change is laced directly into the existing code.  This will allow any further updates to the multi-listener changes to be applied directly to both the epoll and non-epoll versions for comparison testing.  Idle connection checks have been reimplemented using timerfd using epoll as well, eliminating the need for the check_idletimeout function.  I have seen no indication of mysterious performance issues with this patch.

I don’t believe that epoll will provide any significant performance benefit to single connection or even small numbers of connections.  I anticipate that the benefits will start to show with thousands of long-lived connections, like you would see when using 389ds for core name services in a Unix environment, and hopefully at c10k levels.

Firstyear commented on ldap/servers/slapd/daemon.c:
```diff
@@ -775,22 +794,88 @@ disk_monitoring_thread(void *nothing __attribute__((unused)))
+
 static void
+#ifdef ENABLE_EPOLL
+handle_listeners(struct epoll_event *events, int event_count)
```
This function is small enough that you may as well just duplicate it here so that there is a epoll vs old versions of the function. It'll make it a bit clearer to follow.

Firstyear commented on ldap/servers/slapd/daemon.c:
```diff
@@ -775,22 +794,88 @@ disk_monitoring_thread(void *nothing __attribute__((unused)))
             if (SLAPD_POLL_LISTEN_READY(fds[fdidx].out_flags)) {
-                /* accept() the new connection, put it on the active list for handle_pr_read_ready */
+#endif /* ENABLE_EPOLL */
+                /* Question: If the listenfd is non-blocking, shouldn't we loop here until
```
the return here can also be -1 if PR_Accept fails, so we don't want to loop here if that already happened.

Generally if we get to this situation we really are at an edge case where we have filled the connection table, so erroring faster is better than looping, because if we loop then we also cause other connections to be delayed in finding out they can't be accepted.

Firstyear commented on ldap/servers/slapd/daemon.c:
```diff
@@ -826,7 +916,19 @@ accept_thread(void *vports)
     i_unix = ports->i_socket;
 #endif /* ENABLE_LDAPI */
 
+#ifdef ENABLE_EPOLL
```
Similar to the above, I think shortterm duplicating the accept_threads() fn here is better than trying to ifdef so much which makes it harder to reader.

llile replied:
Accepted as above.

Firstyear commented on ldap/servers/slapd/daemon.c:
```diff
@@ -2030,6 +2262,48 @@ handle_new_connection(Connection_Table *ct, int tcps, PRFileDesc *listenfd, int
+    if ((conn->c_idle_tfd = timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK)) < 0) {
+        slapi_log_err(SLAPI_LOG_ERR, "handle_new_connection", "timerfd_create() failed: %s\n",
+                      strerror(errno));
+        /* XXX cleanup how? */
```
If we can't create this, or add to epoll, I think we go straight into disconnect_server_nomutex() so that we can disconnect the connection and log an appropriate error. Generally both of these will error if we are out of fd's or some other resource limit so I think disconnecting and making noise is correct here.

Firstyear replied:
I think you're right. Up until the connection is added to the active list, it seems there is a bit of an odd limbo state here where we don't have a clear way to return this to the freelist on the conntable. I suspect we need to break `connection_table_move_connection_out_of_active_list`. Maybe the code after https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/conntable.c#L467 needs to become a separate function where we can return the connection from these early error cases where disconnect server mutex isn't the right answer, and before we are on the active list.

Really, the active list and the concept of it isn't needed once we have epoll, but that's a bigger cleanup and rework.

Firstyear commented on ldap/servers/slapd/daemon.c:
```diff
@@ -2620,6 +2894,13 @@ get_connection_table_size(void)
     }
 
+#ifdef ENABLE_EPOLL
+    /* epoll and timerfd use 2 descriptors per connection
```
I think the way this is written is a bit unclear, but also we actually need more file descriptors than this. 

We have an epoll fds in the accept threads, and we also have epoll fds in all the handlers too. So I actually think this shourld be:

```
size = (size / 2) - num_listeners - 3 ;
size = size - ((size / 2) % 2);
```

I think that covers it?

llile replied:
After reading your suggestion, does this comment make clearer the calculation?  I have also updated the calculation to hopefully be more accurate, I was originally trying to modulus to the multiple of the number of connection table lists and failed to use the actual count.

    /* We need to account for the additional timer file descriptor per connection,
     * which is used for the idle timeout, the listener file descriptors, the epoll
     * file descriptors in the accept threads and connection table.
     * 
     * Then round the size down to the nearest multiple of the number of connection
     * table lists, so that we can use the same number of connections per list.
     */
    size = size - the_connection_table->list_num - num_listeners;
    size = (size / 2) - ((size / 2) % the_connection_table->list_num);

Firstyear reviewed (commented):
This looks really good so far! I look forward to giving it a test run :)

llile committed:
Migrate from PR_Poll to epoll and timerfd.

llile committed:
Description:

Migrate from PR_Poll to epoll and timerfd for connection
timeout management.

Relates: https://github.com/389ds/389-ds-base/issues/5181

Author: Larry Lile

Reviewed by: firstyear, ???

Firstyear commented on ldap/servers/slapd/connection.c:
```diff
@@ -189,6 +189,14 @@ connection_cleanup(Connection *conn)
     if (conn->c_prfd) {
         PR_Close(conn->c_prfd);
     }
+#ifdef ENABLE_EPOLL_NOTYET
```
Just curious why this is _NOTYET? What was your concern about it?

llile replied:
No concern.  I just failed to revert that when diagnosing an issue.

Firstyear commented on ldap/servers/slapd/connection.c:
```diff
@@ -1938,6 +1958,18 @@ connection_threadmain(void *arg)
+                        /* Reset the idle timer */
+                        slapi_log_err(SLAPI_LOG_DEBUG,
+                                      "handle_pr_read_ready", "resetting idle timer for connection %d to %d\n", conn->c_ci, conn->c_idletimeout);
+                        timerfd_settime(conn->c_idle_tfd, 0,
```
If the connection is in CONN_FLAG_CLOSING, then won't this timerfd_settime fail as the event was removed?

llile replied:
The itimerspec is attached to the actual fd, not the epoll event struct.  It is reset here to avoid firing while closing.

Firstyear commented on ldap/servers/slapd/daemon.c:
```diff
@@ -775,6 +794,83 @@ disk_monitoring_thread(void *nothing __attribute__((unused)))
+    size_t idx;
+    int ctlist = 0;
+    struct listener_info *listener = NULL;
+    for (idx = 0; idx < (size_t)event_count; idx++) {
```
minor, `for (size_t idx = 0; ... )` There are some compiler optimisations that only work if you put it into the for loop like this, and it's a nice defence against using the idx outside the loop generally.

llile committed:
Migrate from PR_Poll to epoll and timerfd
    Description:

    Migrate from PR_Poll to epoll and timerfd for connection
    timeout management.

    Relates: https://github.com/389ds/389-ds-base/issues/5181

    Author: Larry Lile

    Reviewed by: firstyear, ???

llile committed:
Migrate from PR_Poll to epoll and timerfd
        Description:
        Migrate from PR_Poll to epoll and timerfd for connection
        timeout management.

        Relates: https://github.com/389ds/389-ds-base/issues/5181

        Author: Larry Lile

        Reviewed by: firstyear, ???

llile committed:
Migrate from PR_Poll to epoll and timerfd
            Description:
            Migrate from PR_Poll to epoll and timerfd for connection
            timeout management.

            Relates: https://github.com/389ds/389-ds-base/issues/5181

            Author: Larry Lile

            Reviewed by: firstyear, ???

Firstyear reviewed (approved):
This looks great I'm really happy with this. I'll leave it for a day or so for other team members if they want to review, but I'm excited to see how this performs. Thanks so much for the work on this!

Firstyear commented:
@jchapma Since you looked at the connection code recently, maybe you want to look at this too.

jchapma commented:
Hi @llile, its great to see your epoll enhancement back to life. I am reviewing the code and running some performance tests in parallel, I will report my findings on Monday. Looks positive so far. Thank you for your work.

llile commented:
@jchapma thanks for running your performance tests on the epoll variant.  I'm sad to see it doesn't really offer any base level benefit, I'm hoping it will have greater effect under higher connection counts.  

A couple of quick questions: 
1. Which version of slamd are you using?  The Oracle 2.1.0 or Github version?  
2. Do you happen to have your slamd configuration documented anywhere I could access?
3. Do you think slapd_accept_wakeup_timer should be increased or decreased?

llile committed:
Migrate from PR_Poll to epoll and timerfd
Description:
Migrate from PR_Poll to epoll and timerfd for connection
timeout management.

Relates: https://github.com/389ds/389-ds-base/issues/5181

Author: Larry Lile

Reviewed by: firstyear, ???

progier389 reviewed (commented):
As most of the code is within #ifdef.
I think that once you change the configure.ac to disable the feature by default
we can (and should) merge the PR (as @firstyear said, we can always work on the performance later on)
But we cannot enable it by default right now. Apparently it changes the server dynamics (and that is probably why some tests fails ) The tests spend more time than usual  and some tests fails because replication does not synchronize in time 

Many tests are still failing compared to the usual run (and I feel that the tests are much longer . 
apparently replication does not seems to synchronize soon enough - maybe a change of dynamic and a timeout issue)

llile committed:
Migrate from PR_Poll to epoll and timerfd
Description:

Migrate from PR_Poll to epoll and timerfd for connection
timeout management.

Relates: https://github.com/389ds/389-ds-base/issues/5181

Author: Larry Lile

Reviewed by: firstyear, ???

llile commented on configure.ac:
```diff
@@ -395,6 +395,23 @@ else
 AM_CONDITIONAL(enable_posix_winsync,test "$enable_posix_winsync" = "yes")
 
+if test -z "$enable_epoll" ; then
+  enable_epoll=yes
```
I have updated configure.ac in the patch as requested.

tbordaz reviewed (approved):
Looks very nice to me as well. Few remarks. I agree it can be pushed upstream now and improved later.

progier389 reviewed (approved):
As excepted, with the feature disabled the tests results are back to the expected ones. 
Waiting for the changes related to @tbordaz comments before merging the PR

llile commented:
Should I squash the commit?

llile committed:
Migrate from PR_Poll to epoll and timerfd
Description:
Migrate from PR_Poll to epoll and timerfd for connection
timeout management.

Relates: https://github.com/389ds/389-ds-base/issues/5181

Author: Larry Lile

Reviewed by: firstyear, progier, tbordaz

Firstyear commented:
With 4 acks, and consensus we want to merge this, I've gone ahead and done so. Thank you so much @llile for this, we look forward to continuing to improve this feature and it's performance!

llile commented:
Thanks everyone for the assistance with the patch, I hope it will be valuable going forward.
